share/doc/wake/logging.md

```diff
@@ -0,0 +1,111 @@
+Check out [print.wake](https://github.com/sifive/wake/blob/master/share/wake/lib/core/print.wake)
+for commented code.
+
+## Loggers
+
+Wake has a concept of loggers,
+to which bytes can be written during wake execution by various sources.
+Based on command line flags, each logger's output will be directed to
+zero or more of `wake`'s own output streams, possibly with additional formatting applied.
+The wake codebase and standard libraries often refer to a _logger_ as a `logLevel`.
+
+What's the difference between the different loggers?
+Which of wake's own output streams the results are sent to.
+This can influence what you see on your console when you run wake.
+The logger, in combination with the command line flags you invoke wake with,
+controls *what* you see. The logger also controls what it looks like (what color it is).
+It also controls which *actual* stream it is going to coming out of the executing `wake` process
+(stderr or stdout, or other file descriptors 3-5).
+
+`wake` has a number of built in loggers in wake which all start with `log...`.
+You can also make your own logger using `mkLogLevel`.
+You can use this to customize the color you want it to be shown as (if at all),
+and giving more fine-grained control over which wake output stream it should go to.
+
+For stdout, the following table should be read top to bottom considering the command line arguments you've passed to wake.
+If you match a line in the table, stop!
+This is what will appear on the console as you watch wake execute, and what is sent to wake's stdout:
+
+|Flag               | logDebug | logInfo | logEcho | logReport | logWarning | logError | logNever | mkLogLevel "foo" Green|
+|-------------------|----------|---------|---------|-----------|------------|----------|----------|----------------------
+|Color on Console   |  Blue    |  Dim    |  default| Magenta   |  Yellow    |   Red    |          |      Green      |          
+|--stdout="foo"     |          |         |         |           |            |          |          |         x       |
+|--stdout="foo,info" |         |    x    |         |           |            |          |          |         x       |
+|--debug            |      x   |    x    |    x    |     x     |      x     |    x     |          |                 |
+|--verbose          |          |    x    |    x    |     x     |      x     |    x     |          |                 |
+|--quiet            |          |         |         |           |            |    x     |          |                 |
+|--no-tty<sup>*</sup>|         |         |    x    |     x     |      x     |    x     |          |                 |
+|(default)          |          |         |         |     x     |      x     |    x     |          |                 |
+
+<sup>*</sup> `--no-tty` also prevents any coloring from being applied.
+
+
+Unless overridden with `--stderr=...` the *only* thing that ever appears on stderr coming out of wake is `logError`,
```

Comment:
Once again, you are conflating the name of the variable with the name of the logger. You should call it "error" (or probably the "error logger").

You can have multiple variables point to the same object in memory, so the only canonical name will be the name of the actual logger.

Reply:
yeah now I am curious what actually happens if I `mkLogLevel "mega"` in two places... I assume that both created loggers would still be impacted by e.g. `--stderr=mega`

Reply:
I'm not quite sure what you are suggesting. I *think* you saying that the source of truth of the logger is the string itself, so "error" or "mega", and `mkLogLevel` just makes something that will direct to that source-of-truth logger? But I think that's not quite true, because I can `mkLogLevel "mega" Green` and `mkLogLevel "mega" Blue` and those are different loggers, coupling the level and the color displayed...

Reply:
let me re-read what I've written and your comments and try to be more consistent. I want to use these definitions:

* log level is something like "error", "info", "foo"
* a logger couples a log level and a color
* command line arguments influence what happens to things sent to loggers of a given log level
* outputs are directed to a logger

WDYT?

Reply:
> I'm not quite sure what you are suggesting. I think you saying that the source of truth of the logger is the string itself, so "error" or "mega", and mkLogLevel just makes something that will direct to that source-of-truth logger? But I think that's not quite true, because I can mkLogLevel "mega" Green and mkLogLevel "mega" Blue and those are different loggers, coupling the level and the color displayed...

No, I'm saying in

```wake
def foo = mkLogLevel "mega" Green
def bar = foo
```

`foo` and `bar` are wake variable names that point to the same object in memory, which is a logger by the name of "mega". I'm saying it's incorrect or at least a little bit confusing for you to refer to the loggers by their wake variable names rather than by the value of the `name` field on these objects.

Reply:
right, i think this is in aligment with what I am saying... the strings are really their log level and what the command lines are dealing with. There may be many loggers out there (by whatever wake variable name) with the same stringy log level.

Reply:
I think some of my misunderstanding is that you're trying to use terminology that, although makes more sense, is inconsistent with the wake code code.

> log level is something like "error", "info", "foo"
> a logger couples a log level and a color

The confusing thing is that the wake code calls the second thing a `LogLevel` and the first thing a `name`.

Reply:
I added a table to distinguish terms used in this doc vs what wake calls them internally, then tried to use the former correctly and consistently throughout the doc.